overlord/snapshotstate/backend/backend.go

```diff
@@ -0,0 +1,256 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package backend
+
+import (
+	"archive/zip"
+	"crypto"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+const (
+	archiveName  = "archive.tgz"
+	metadataName = "meta.json"
+	metaHashName = "meta.sha3_384"
+
+	userArchivePrefix = "user/"
+	userArchiveSuffix = ".tgz"
+)
+
+var (
+	dirOpen  = os.Open
+	dirNames = (*os.File).Readdirnames
+	open     = Open
+)
+
+// Iter loops over all snapshots in the snapshots directory, applying the
+// given function to each. The snapshot will be closed after the function
+// returns. If the function returns error, iteration is stopped (and if the
+// error isn't EOF, it's returned as the error of the iterator).
+func Iter(ctx context.Context, f func(*Reader) error) error {
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	dir, err := dirOpen(dirs.SnapshotsDir)
+	if err != nil {
+		if osutil.IsDirNotExist(err) {
+			// no dir -> no snapshots
+			return nil
+		}
+		return fmt.Errorf("cannot open snapshots directory: %v", err)
+	}
+	defer dir.Close()
+
+	var names []string
+	for err == nil {
+		names, err = dirNames(dir, 100)
+		for _, name := range names {
+			if err = ctx.Err(); err != nil {
+				break
+			}
+
+			filename := filepath.Join(dirs.SnapshotsDir, name)
+			rsh, openError := open(filename)
+			// rsh can be non-nil even when openError is not nil (in
+			// which case rsh.Broken will have a reason). f can
+			// check and either ignore or return an error when
+			// finding a broken snapshot.
+			if rsh != nil {
+				err = f(rsh)
+			} else {
+				// TODO: use warnings instead
+				logger.Noticef("cannot open snapshot %q: %v", name, openError)
+			}
+			if openError == nil {
+				// if openError was nil the snapshot was opened and needs closing
+				if closeError := rsh.Close(); err == nil {
+					err = closeError
+				}
+			}
+			if err != nil {
+				break
+			}
+		}
+	}
+
+	if err == io.EOF {
```

Comment:
The wording on Readdirnames makes me think that it might at some point return a non-empty list and a non-nil error: it's documented that _if_ names is empty, then the error is non-nil. It does not say that a non-nil error implies names is empty.

The implementation confirms this suspicion.

I'll add a comment explaining this.